pkg/core/xds/rules.go

```diff
@@ -0,0 +1,236 @@
+// If filter out rules with negative tags the order becomes from the most specific to the less
+// specific rules. Filtering out of negative rules could be useful for XDS generators that
+// don't have a way to configure negations.
+//
+// See the detailed algorithm description in docs/madr/decisions/007-mesh-traffic-permission.md
+func BuildRules(list []PolicyItem) Rules {
+	rules := Rules{}
+
+	// 1. Each targetRef should be represented as a list of tags
+	tagSet := map[Tag]bool{}
+	for _, item := range list {
+		for _, t := range asSubset(item.GetTargetRef()) {
+			tagSet[*t] = true
+		}
+	}
+	tags := []*Tag{}
+	for tag := range tagSet {
+		v := tag
+		tags = append(tags, &v)
+	}
+
+	sort.Slice(tags, func(i, j int) bool {
+		if tags[i].Key != tags[j].Key {
+			return tags[i].Key < tags[j].Key
+		}
+		return tags[i].Value < tags[j].Value
+	})
+
+	// 2. Iterate over all possible combinations with negations
+	iter := NewSubsetIter(tags)
+	for {
+		ss, next := iter.Next()
+		// 3. For each combination determine a configuration
+		if ss != nil {
+			var conf proto.Message
+			for i := 0; i < len(list); i++ {
+				item := list[i]
+				if asSubset(item.GetTargetRef()).IsSubset(ss) {
+					if conf == nil {
+						conf = proto.Clone(item.GetDefaultAsProto())
+					} else {
+						// todo(lobkovilya): util_proto.Merge appends lists,
+						// create a custom Merge func for list replacements
+						util_proto.Merge(conf, item.GetDefaultAsProto())
+					}
+				}
+			}
+			rules = append(rules, &Rule{
+				Subset: ss,
+				Conf:   conf,
+			})
+		}
+		if !next {
+			break
+		}
+	}
+
+	sort.Slice(rules, func(i, j int) bool {
+		return rules[i].Subset.NumPositive() > rules[j].Subset.NumPositive()
+	})
+
+	return rules
+}
+
+func asSubset(tr *common_api.TargetRef) Subset {
+	if tr == nil {
+		// syntactic sugar, empty targetRef means targetRef{kind: Mesh}
+		return Subset{}
+	}
+	switch tr.GetKindEnum() {
+	case common_api.TargetRef_Mesh:
+		return Subset{}
+	case common_api.TargetRef_MeshSubset:
+		ss := Subset{}
+		for k, v := range tr.GetTags() {
+			ss = append(ss, &Tag{Key: k, Value: v})
+		}
+		return ss
+	case common_api.TargetRef_MeshService:
+		return Subset{{Key: mesh_proto.ServiceTag, Value: tr.GetName()}}
+	case common_api.TargetRef_MeshServiceSubset:
+		ss := Subset{{Key: mesh_proto.ServiceTag, Value: tr.GetName()}}
+		for k, v := range tr.GetTags() {
+			ss = append(ss, &Tag{Key: k, Value: v})
+		}
+		return ss
+	default:
+		panic("unsupported type to represent as tags")
+	}
+}
+
+type SubsetIter struct {
+	current Subset
+}
+
+func NewSubsetIter(ss Subset) *SubsetIter {
+	return &SubsetIter{
+		current: ss,
+	}
+}
+
+func (c *SubsetIter) Next() ([]*Tag, bool) {
+	for idx := 0; idx < len(c.current); idx++ {
+		if c.current[idx].Not {
+			c.current[idx].Not = false
+			continue
+		} else {
+			c.current[idx].Not = true
+			return c.simplified(), true
+		}
+	}
+
+	return c.simplified(), false
+}
+
+func (c *SubsetIter) simplified() Subset {
```

Comment:
I'm having a hard time to get what we are doing here. Any short description?

Reply:
added a comment for this method